0005-ordered-quantification.md

```diff
@@ -0,0 +1,644 @@
+
+```dafny
+datatype Suit = Clubs | Diamonds | Hearts | Spades
+
+foreach s: Suit {
+  ...
+}
+
+foreach b1: bool, b2: bool {
+  expect TestMyFunctionWithOptions(b1, b2) == true;
+}
+
+type SmallIndex = x: nat | 0 <= x < 8
+
+foreach i: SmallIndex {
+  ...
+}
+```
+
+There will not initially (see Open Questions) be any builtin support for automatically tracking the enumerated values so far, 
+as there is for `iterator` values (see https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-iterator-types). 
+It is straightforward to track these values manually, however:
+
+```dafny
+ghost var xs := [];
+foreach x <- s {
+  ...
+  xs := xs + [x];
+}
+```
+
+It is even possible to attach this ghost state directly to the sequence with a helper function:
+
+```dafny
+foreach xAndXs <- WithPrefixes(c) {
+  var (x, xs) := xAndXs;
+  ...
+}
+```
+
+Where the signature of `WithPrefixes` is something like:
+
+```dafny
+function method WithPrefixes<T>(s: seq<T>): seq<(T, ghost seq<T>)> {
+  ...
+}
+```
+
+Similarly, a helper function can be provided to maintain a running index of the enumerated values:
+
+```dafny
+foreach xAndI <- WithIndexes(s) {
+  var (x, i) := xAndI;
+  ...
+}
+
+// Alternatively, if s is a sequence:
+foreach i, x | 0 <= i < |s| && s[i] == x {
+  ...
+}
+```
+
+## Sequence comprehensions
+
+A sequence comprehension has identical syntax to a set comprehension, except that it begins with
+`seq` rather than `set`. Therefore its syntax is:
+
+```
+SeqComprehension ::=
+  "seq" 
+  QuantifierVarDecl { "," QuantifierVarDecl } 
+  "|" Expression
+  [ :: Expression ]
+```
+
+Sequence comprehensions are likely to be easier to work with than `foreach` loops, and Dafny users should be
+encouraged to use them instead where possible. Many high-order operations on sequences can be expressed directly
+using sequence comprehensions:
+
+```dafny
+// Filtering to optional values
+var s: seq<Option<T>> := ...;
+var filtered: seq<T> := seq o <- s | o.Some? :: o.value;
+
+// Zipping two lists together
+var a: seq<T> := ...;
+var b: seq<T> := ...;
+assert |a| == |b|;
+var zipped := seq i | 0 <= i < |a| :: (a[i], b[i]);
+
+// Map and then flatten (a.k.a. "FlatMap")
+var c: seq<S> := ...;
+var f: S -> seq<T> := ...;
+var flatMapped: seq<T> := seq x <- c, y <- f(x) :: y;
+
+// Sorted sequence from a set
+var setOfReals := {3.141, 2.718, 1.618};
+var sortedList := seq x | x in setOfReals;
+assert sortedList == {1.618, 2.718, 3.141};
+
+// Intersection of sequences
+var a := [4, 1, 3, 5];
+var b := [3, 1, 7];
+assert (seq x <- a | x in b) == [1, 3];
+assert (seq x <- b | x in a) == [3, 1];
+```
+
+Since sequence comprehensions are expressions rather than statements, they carry an additional restriction
+when compiled: their ordering must be fully deterministic. If `s` is a `set<int>`, for example,
+`seq x <- s` would be rejected in compiled code, whereas `seq x | x in s` would be allowed.
```

Comment:
> I'm not sure I would buy into that. Deterministic means that the result will always be the same if the same set is given. The compilers should allow traversing a set in a deterministic order if they are equal anyway, even if it just means comparing references of non-comparable types. So `seq x <- s` could be accepted in compiled code.

This can be done in some cases but definitely makes the runtime more complicated: most HashSet implementations might give you a different iteration order for `{1, 2, 3}` vs. `{3, 2, 1}`. I tried to explain how that could be addressed in many cases with a more specialized implementation in [this section in the original RFC](https://github.com/dafny-lang/rfcs/pull/9/files#diff-58bd66111c2a9e8e41d93574b6b89dd40f268d2c802943c9a3fe0380f853ac9dR413), but didn't go back to clean that up since I've dropped the requirement in this version. I can take another stab at it here if it would help though.

> I suspect that, most of the time, however, users will not want to convert sets to sequences, but always use sequences, and use set in ghost contexts.

Well now that it's public (🎉!), allow me to point to the natively-implemented [`ComputeSetToOrderedSequence`](https://github.com/aws/aws-encryption-sdk-dafny/blob/mainline/src/Util/Sets.dfy#L11) method that the ESDK uses in several places to deterministically serialize the keys of a map!